infrastructure/src/main/java/org/corfudb/infrastructure/CompactorService.java

```diff
@@ -0,0 +1,163 @@
+                     @NonNull ICompactionTriggerPolicy compactionTriggerPolicy) {
+        this.serverContext = serverContext;
+        this.runtimeSingletonResource = runtimeSingletonResource;
+
+        String threadName;
+        if (serverContext.getServerConfig().get("<port>").toString().equals("9040")) {
+            threadName = "Cmpt-NonConfig-chkpter";
+        } else {
+            threadName = "Cmpt-Config-chkpter";
+        }
+        this.orchestratorThread = Executors.newSingleThreadScheduledExecutor(
+                new ThreadFactoryBuilder()
+                        .setNameFormat(threadName)
+                        .build());
+        this.checkpointerJvmManager = checkpointerJvmManager;
+        this.compactionTriggerPolicy = compactionTriggerPolicy;
+        syslog = LoggerFactory.getLogger("SYSLOG");
+    }
+
+    CorfuRuntime getCorfuRuntime() {
+        return runtimeSingletonResource.get();
+    }
+
+    /**
+     * Starts the long running service.
+     *
+     * @param interval interval to run the service
+     */
+    @Override
+    public void start(Duration interval) {
+        this.compactorLeaderServices = new CompactorLeaderServices(getCorfuRuntime(), serverContext.getLocalEndpoint());
+        this.corfuStore = new CorfuStore(getCorfuRuntime());
+        this.compactionTriggerPolicy.setCorfuRuntime(getCorfuRuntime());
+        if (getCorfuRuntime().getParameters().getCheckpointTriggerFreqMillis() > 0) {
+            this.compactionTriggerFreqMs = getCorfuRuntime().getParameters().getCheckpointTriggerFreqMillis();
+        }
+
+        orchestratorThread.scheduleAtFixedRate(
+            this::runOrchestrator,
+            interval.toMillis(),
+            interval.toMillis(),
+            TimeUnit.MILLISECONDS
+        );
+    }
+
+    /**
+     * Invokes and cleans up the CorfuStoreCompactor jvm based on the status of CompactionManager
+     * Additionally, If the current node is the leader,
+     *      a. Invokes ValidateLiveness() to keep track of checkpointing progress by each client
+     *      b. Triggers the distributed compaction cycle based on the TriggerPolicy
+     */
+    private void runOrchestrator() {
+        boolean isLeader = isNodePrimarySequencer(updateLayoutAndGet());
+        compactorLeaderServices.setLeader(isLeader);
+        try (TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE)) {
+            CheckpointingStatus managerStatus = (CheckpointingStatus) txn.getRecord(
+                    DistributedCompactor.COMPACTION_MANAGER_TABLE_NAME,
+                    DistributedCompactor.COMPACTION_MANAGER_KEY).getPayload();
+            txn.commit();
+            syslog.trace("ManagerStatus: {}", managerStatus == null ? "null" : managerStatus.getStatus());
+            if (managerStatus != null) {
+                if (managerStatus.getStatus() == StatusType.FAILED || managerStatus.getStatus() == StatusType.COMPLETED) {
```

Comment:
How it can be possible that a task Completed and Failed at the same time?

Reply:
It cannot be. We want to enter the if clause if one of them is true.